pep-0718.rst

```diff
@@ -0,0 +1,152 @@
+PEP: 718
+Title: Subscriptable functions
+Author: James Hilton-Balfe <gobot1234yt@gmail.com>
+Sponsor: Guido van Rossum <guido@python.org>
+Discussions-To: https://discuss.python.org/t/26463/
+Status: Draft
+Type: Standards Track
+Topic: Typing
+Content-Type: text/x-rst
+Created: 23-Jun-2023
+Python-Version: 3.13
+Post-History: `24-Jun-2023 <https://discuss.python.org/t/28457/>`__
+
+Abstract
+--------
+
+This PEP proposes making ``FunctionType``\ , ``MethodType``\ , ``BuiltinFunctionType``
+and ``BuiltinMethodType`` instances subscriptable for typing purposes. Doing so allows
+for specialisation in complex, hard-to-infer situations for type checkers where
+bi-directional inference (which allows for the types of parameters of anonymous
+functions to be inferred) is insufficient.
+
+Motivation
+----------
+
+Currently, it is not possible to infer certain situations for generic functions where
```

Comment:
When I'm saying cannot infer in this PEP I mean infer a type more specific than Any/Unknown because these in my mind are not "proper", "useful" types. Also FWIW bounds don't play any part in the type that's picked for the return type 
![image](https://github.com/python/peps/assets/50501825/a59886e0-8279-467d-97a1-5ba28a3503af)
This is something that default types for type parameters (PEP 696) could maybe solve if it was extended to functions, which after this PEP maybe it could be, using reified types.